examples/mnist/train_mnist.py

```diff
@@ -77,7 +77,9 @@ def main():
     trainer = training.Trainer(updater, (args.epoch, 'epoch'), out=args.out)
 
     # Evaluate the model with the test dataset for each epoch
-    trainer.extend(extensions.Evaluator(test_iter, model, device=args.gpu))
+    evaluator = extensions.Evaluator(test_iter, model, device=args.gpu)
+    evaluator.initialize = evaluator.__call__  # for evaluation at epoch 0
```

Comment:
I am not sure about this interface. Can we not add an argument to the constructor instead?

Reply:
I think there are 3 options:

1. `invoke_before_training=False/True` which implicitly overwrites `initialize` by `__call__`. This can cause unexpected behavior if user overwrites `Evaluator` and defines a custom `initialize` in it.
2. Passing `initialize=None/Extension.__call__/etc.` to the constructor and it overwrites the `initialize` method of the Extension if it is not `None`. This is fine, but it might be confusing that there is no way to overwrite other methods for example `__call__` and `finalize`. Also, it can't overwrite `initialize` method of extensions created by `extensions.make_extension`, for example `extensions.snapshot()`, so user must manually overwrite it.
3. Current implementation. It explicitly overwrites `initialize`, so it is less confusing. 

Currently, I'm conservative to adopt other implementations because of above problems caused by overwriting method in `__init__`.
What do you think?

Reply:
I think 1. is best but with some slight modifications. What do you think of the following?

- `call_before_training=False/True` passed to both `extend` and the constructors of the extensions. This flag is checked to decide whether to fire or not.
- Do not change the behavior of `initialize` since it's actual designed for a different purpose. Namely to update the trainer on resume.
- Make sure that the "0 iteration" invocations is not triggered on resume (check the iteration count).